File: TwoNumberSum.java
/*

Question 1: Two Number Sum

Given an array of integers nums and an integer target, return the
the two elements such that their sum equals target. You may assume
that exactly one solution exists, and you may not use the same
element twice.

Example:
nums = [2, 7, 11, 15]
target = 9

Output: [2, 7]

Explanation: nums[0] + nums[1] = 2 + 7 = 9
*/

import java.util.Arrays;
import java.util.HashMap;

public class TwoNumberSum {

    // Solution One
    // This is brutal force solution using two for loops
    // O(n^2) time | O(1) space
    public static int[] solution1(int[] nums, int target) {
        for (int i = 0; i < nums.length; i++) {
            for (int j = i; j < nums.length; j++) {
                int firstNumber = nums[i];
                int secondNumber = nums[j];
                int sum = firstNumber + secondNumber;
                if (sum == target) {
                    return new int[]{firstNumber, secondNumber};
                }
            }
        }
        return new int[]{};
    }

    // Solution Two
    // Using hashmap(dict) to store the value and find the pair
    // O(n) time | O(n) space

    public static int[] solution2(int[] nums, int target) {
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            int currentNumber = nums[i];
            Integer pairNumber = map.get(currentNumber);
            if (pairNumber != null) {
                return new int[]{pairNumber, currentNumber};
            }
            map.put(target - currentNumber, currentNumber);
        }
        return new int[]{};
    }

    // Solution Three
    // Using sort and two pointer to find the pair
    // O(nlog(n)) time | O(1) space

    public static int[] solution3(int[] nums, int target) {
        Arrays.sort(nums);
        int left = 0;
        int right = nums.length - 1;
        while (left < right) {
            int firstNumber = nums[left];
            int secondNumber = nums[right];
            int currentSum = firstNumber + secondNumber;
            if (currentSum > target) {
                right--;
            } else if (currentSum < target) {
                left++;
            } else {
                return new int[]{firstNumber, secondNumber};
            }
        }
        return new int[]{};
    }
}

File: junior level/TwoNumberSum.py
"""
Two Number Sum

Given an array of integers nums and an integer target, return the
the two elements such that their sum equals target. You may assume
that exactly one solution exists, and you may not use the same
element twice.

Input: [2, 7, 11, 15], 9

Output: [2, 7]

"""

# O(n^2) time | O(1) space
def solution1(array, target):
    for i in range(len(array) - 1):
        firstNumber = array[i]
        for j in range(i + 1, len(array)):
            secondNumber = array[j]
            if firstNumber + secondNumber == target:
                return [firstNumber, secondNumber]
    return []

# O(n) time | O(n) space
def solution2(array, target):
    nums = {}
    for num in array:
        if (target - num) in nums:
            return [target - num, num]
        else:
            nums[num] = True
    return []

# O(nlogn) time | O(1) space
def solution3(array, target):
    array.sort()
    left = 0
    right = len(array) - 1
    while left < right:
        sum = array[left] + array[right]
        if sum == target:
            return [array[left], array[right]]
        elif sum > target:
            right -= 1
        else:
            left += 1
    return []

print(solution1([2, 7, 11, 15], 9))
print(solution2([2, 7, 11, 15], 9))
print(solution3([2, 7, 11, 15], 9))
